### PR TITLE
feat: add list collection

### PR DESCRIPTION
**Pull request by anitarua**
Addresses #165 

Prioritized getting these APIs implemented:

- listConcatenateBack
- listConcatenateFront
- listFetch
- listLength
- listPopBack
- listPopFront
- listRemoveValue

listPushFront and listPushBack were also implemented in the simple_cache_client but seemed less urgent to include them here since the same functionality can be done using the concatenate methods.

### CODE REVIEW

File: src/cache/requests/list/list_fetch.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            .into_inner();
+
+        match response.list {
+            Some(list_fetch_response::List::Missing(_)) => Ok(ListFetch::Miss),
+            Some(list_fetch_response::List::Found(found)) => Ok(ListFetch::Hit {
+                values: ListFetchValue {
+                    raw_item: found.values,
+                },
+            }),
+            _ => unreachable!(),
```
Comment (malandis) [inline src/cache/requests/list/list_fetch.rs:113]: In the sorted set implementation we had a case for None, which we interpreted as a `Miss`

File: src/cache/requests/list/list_concatenate_back.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
+use crate::{
+    cache::{CollectionTtl, MomentoRequest},
+    utils::prep_request_with_timeout,
+    CacheClient, IntoBytes, MomentoResult,
+};
+
+/// Adds multiple elements to the back of the given list. Creates the list if it does not already exist.
```
Comment (malandis) [inline src/cache/requests/list/list_concatenate_back.rs:7]: I would add a remark that the order the new values appear in the list is identical to the order of the values passed in.

File: src/cache/requests/list/list_concatenate_front.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
+use crate::{
+    cache::{CollectionTtl, MomentoRequest},
+    utils::prep_request_with_timeout,
+    CacheClient, IntoBytes, MomentoResult,
+};
+
+/// Adds multiple elements to the front of the given list. Creates the list if it does not already exist.
```
Comment (malandis) [inline src/cache/requests/list/list_concatenate_front.rs:7]: Same remark about order applies here.

File: src/cache/requests/list/list_fetch.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+/// Gets a list item from a cache with optional slices.
+///
+/// # Arguments
+/// * `cache_name` - name of cache
+/// * `list_name` - name of the list
+///
+/// # Optional Arguments
+///
+/// * `start_index` - The starting inclusive element of the list to fetch. Default is 0.
+/// * `end_index` - The ending exclusive element of the list to fetch. Default is end of list.
```
Comment (malandis) [inline src/cache/requests/list/list_fetch.rs:26]:
```suggestion
/// * `end_index` - The ending exclusive element of the list to fetch. Default is up to and including end of list.
```

**Review by malandis — COMMENTED** [review]
Few comments about documentation.

File: src/cache/requests/list/list_fetch.rs (hunk 1 above)
Comment (anitarua) [reply src/cache/requests/list/list_fetch.rs:113]: I could return an unknown error instead like so? (and update accordingly for other instances as well)

```
MomentoError {
    message: "Unknown error has occurred, unable to parse list_fetch_response".into(),
    error_code: MomentoErrorCode::UnknownError,
    inner_error: Some(response),
    details: None,
}
```

Comment (malandis) [reply src/cache/requests/list/list_fetch.rs:113]: Yes that looks good. That follows the JS pattern [here](https://github.com/momentohq/client-sdk-javascript/blob/b0ed1b3ac745ee1eeca1cc8cbc672b1a151268cc/packages/client-sdk-nodejs/src/internal/cache-data-client.ts#L2505-L2515) and [here](https://github.com/momentohq/client-sdk-javascript/blob/main/packages/client-sdk-nodejs/src/errors/cache-service-error-mapper.ts#L99). We can extract this to a function later.

Comment (anitarua) [reply src/cache/requests/list/list_fetch.rs:113]: Made a ticket to do that throughout the sdk: https://github.com/momentohq/client-sdk-rust/issues/239

**Review by malandis — APPROVED** [review]

**Pull request merged**